zip-0221.rst

```diff
@@ -0,0 +1,743 @@
+root ``M_n`` of the MMR MUST be included in the header of ``B_n``.
+
+(``x`` is the activation height of the most recent upgrade network upgrade)
+
+FlyClient reduces the number of block headers needed for light client verification of a
+valid chain, from linear (as in the current reference protocol) to logarithmic in
+blockchain length. This verification is correct with high probability. It also allows
+creation of subtree proofs, so light clients need only check blocks later than the most
+recently verified block index. Following that, verification of a transaction inclusion
+within that block follows the usual reference protocol [#zip-0307]_.
+
+A smaller proof size could enable the verification of Zcash SPV Proofs in blockchains such
+as Ethereum, enabling efficient cross-chain communication and pegs. It also reduces
+bandwidth and storage requirements for resource-limited clients like mobile or IoT
+devices.
+
+
+Specification
+=============
+
+The leaves of the MMR at block ``B_n`` are hash commitments to the header data and
+metadata of each previous block ``B_x, ..., B_(n-1)``, where ``x`` is the block height of
+the most recent network upgrade. We extend the standard MMR to allow metadata to propagate
+upwards through the tree by either summing the metadata of both children, or inheriting
+the metadata of a specific child as necessary. This allows us to create efficient proofs
+of selected properties of a range of blocks without transmitting the entire range of
+blocks or headers.
+
+Tree Node specification
+-----------------------
+
+Unless otherwise noted, all hashes use BLAKE2b-256 with the personalization field set to
+``'ZcashHistory' || CONSENSUS_BRANCH_ID``. ``CONSENSUS_BRANCH_ID`` is the little-endian
+encoding of ``BRANCH_ID`` for the epoch of the block containing the commitment.
+[#zip-0200]_ Which is to say, each node in the tree commits to the consensus branch that
+produced it.
+
+Each MMR node is defined as follows:
+
+1. ``hashSubtreeCommitment``
+
+- If the node is a leaf node, then ``hashSubtreeCommitment`` is the consensus-defined
+  block hash for the corresponding block.
+
+  * This hash is encoded in internal byte order, and does NOT use the BLAKE2b-256
+    personalization string described above.
+  * For clarity, the ``hashSubtreeCommitment`` field of leaf ``n-1`` is *precisely equal*
+    to the ``hashPrevBlock`` field of header ``n``
+
+- If the node is an internal or root node
+
+  * Both child nodes are serialized
+  * ``hashSubtreeCommitment`` is the BLAKE2b-256 hash of ``left_child || right_child``
+  * For clarity, this digest uses the BLAKE2b-256 personalization string described above.
+- serialized as ``char[32]``
+
+2. ``nEarliestTimestamp``
+
+- If the node is a leaf node
+
+  * ``nEarliestTimestamp`` is the header's timestamp
+
+- If the node is an internal or root node
+
+  * ``nEarliestTimestamp`` is inherited from the left child
+
+- serialized as ``nTime`` (``uint32``)
```

Comment:
Simplicity for now, I guess; by definition this can never contain a value larger than the `nTime` field in the block header, which is a `uint32`. We would change the latter in a network upgrade, and can change this at the same time.